### PR TITLE
Epoch-bind onion exits across process restarts

### LINKED ISSUE #750: Persist or epoch-bind onion replay witnesses across exit restarts
Parent: #743
Severity: High
Review basis: static read-only review of c9d57b581cf1a369327512f2c84fc5ca24a0f482; no local tests/builds were run.

## Finding

The replay invariant is process-local, while the exit decryption key is persistent by default. After an exit process restarts with the same session key, already-consumed forward nonces are forgotten but still-valid old cells can still decrypt.

Evidence:
- the replay cache and replay partitions are in-memory HashMaps with empty defaults: https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/src/onion/replay.rs#L110-L150 and https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/src/onion/replay.rs#L210-L288
- the circuit law says replay witnesses live for the same interval as still-valid forward layers; the max validity is 150 seconds: https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/src/onion/circuit/mod.rs#L78-L87
- TCP exit consumes the nonce before opening the external target, but the witness lives only in OnionTcpRuntime: https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/src/onion/tcp/mod.rs#L267-L305 and https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/src/onion/tcp/mod.rs#L396-L425 and https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/src/onion/tcp/mod.rs#L484-L501 and https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/src/onion/tcp/mod.rs#L522-L560
- HTTPS exit has the same process-local replay state and consumes before the outbound fetch: https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/src/onion/https/mod.rs#L157-L204 and https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/src/onion/https/mod.rs#L616-L648
- native config reads a persisted/raw session_sk, and the CLI default writes ~/.rings/session_sk with a 30-day TTL before passing it into the onion runtime: https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/src/native/config.rs#L156-L168 and https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/src/native/config.rs#L259-L280 and https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/bin/rings.rs#L478-L528 and https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/crates/node/bin/rings.rs#L814-L874
- SECURITY.md describes the intended fail-stop model as restart with a new session, which the native default does not enforce: https://github.com/RingsNetwork/rings/blob/c9d57b581cf1a369327512f2c84fc5ca24a0f482/SECURITY.md#L39-L42

## Concrete argument

A malicious or colluding immediate predecessor records a valid first-hop-to-exit TCP Open or HTTPS POST onion cell. The exit process restarts inside the 150-second forward validity window and reloads the same session key from disk. The attacker then wraps and sends the same onion cell bytes inside a fresh authenticated overlay message. Decryption succeeds, the empty replay cache returns Consumed, and the exit opens the TCP target or repeats the HTTPS request. The response path may fail after restart, but the external side effect has already happened.

## Fix direction

Either persist the one-shot witness set durably until the maximum forward validity window expires, or bind a boot/process epoch into advertised exit descriptors and the forward AEAD transcript so old cells cannot decrypt after restart. Add a regression that constructs two runtimes with the same SessionSk, replays the same valid first forward cell after the first runtime consumes it, and expects a replay rejection or epoch failure before any exit side effect.

### PR DESCRIPTION
**Pull request by zkjoie**
### Please check if the PR fulfills these requirements

- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

## :large_blue_circle: What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)

Security bug fix for onion-exit replay admission across process restarts.

## :brown_circle: What is the current behavior? (You can also link to an open issue here)

Forward replay witnesses are process-local, while a delegated `SessionSk` can persist across restarts. Restarting an exit therefore clears its replay cache and can allow the same still-valid authenticated cell to reach the exit adapter again.

## :green_circle: What is the new behavior (if this is a feature change)?

- Generate a fresh random `OnionExitEpoch` for every processor lifetime.
- Sign that epoch into onion-exit descriptors and authenticate it inside the final forward layer.
- Reject an epoch mismatch in the pure reducer before emitting any exit-adapter effect.
- Carry the epoch through native and browser registration/RPC boundaries.
- Keep epoch-bearing exit registration crate-internal so advertisement and admission cannot be configured with unrelated lifecycle values.
- Add a regression test that reuses the same `SessionSk` and ciphertext with an empty replay cache after restart, then proves that no exit effect is emitted.
- Document the fail-stop restart contract in `SECURITY.md`.

## :radioactive: Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)

Yes, at the pre-v1.0 onion boundary: the exit descriptor schema moves from v2 to v3, the exit descriptor RPC shape gains `process_epoch`, and `OnionExitRegistration` is no longer exported as a public custom-registration task because it cannot safely choose an epoch independently of circuit admission. Stale descriptors fail closed until refresh. The onion AEAD namespace remains v1; no backward-compatibility guarantee is made before v1.0.

## :information_source: Other information

Validated locally with:

- `cargo test -p rings-node --all-targets --all-features`
- `cargo check -p rings-node --all-targets --all-features`
- strict native and wasm/browser Clippy profiles
- `cargo clippy --all --tests -- -D warnings`
- `RUSTDOCFLAGS=-Dwarnings cargo doc --all --no-deps`
- `cargo +nightly fmt --all -- --check`
- `git diff --check`

Closes #750

### CODE REVIEW

File: crates/node/src/onion/circuit/tests/test_circuit_protocol.rs
Comment (RyanKung) [inline crates/node/src/onion/circuit/tests/test_circuit_protocol.rs:1]: I found that almost all tests in this file are mark with `rings_native`, why not just make the mark at `mod.rs`

File: crates/node/src/onion/circuit/tests/test_circuit_protocol.rs [hunk 1]
```diff
@@ -88,6 +97,10 @@ pub(super) fn test_payload(label: &str) -> OnionCircuitPayload {
     )
 }
 
+fn process_epoch() -> OnionExitEpoch {
```
Comment (RyanKung) [inline crates/node/src/onion/circuit/tests/test_circuit_protocol.rs]: Why 17;16, need comment here

File: crates/node/src/onion/circuit/tests/test_circuit_protocol.rs [hunk 2]
```diff
@@ -998,6 +1035,72 @@ async fn test_two_relays_peel_fixed_size_cells_through_the_exit_reducer_and_shel
...
 }
 
+#[cfg(rings_native)]
+#[tokio::test]
+async fn test_restarted_exit_rejects_old_epoch_before_replay_state_or_side_effect() {
+    let client = session();
+    let exit = session();
+    let route = route(&[], &exit);
+    let original_epoch = route.exit().process_epoch;
+    let circuit_id = OnionCircuitId::new([41; 16]);
```
Comment (RyanKung) [inline crates/node/src/onion/circuit/tests/test_circuit_protocol.rs]: Same reason, need comment, why 41;16

File: crates/node/src/onion/circuit/tests/test_circuit_protocol.rs [hunk 3]
```diff
@@ -998,6 +1035,72 @@ async fn test_two_relays_peel_fixed_size_cells_through_the_exit_reducer_and_shel
     assert_eq!(second_transition.state.relay_return_count(), 1);
 }
 
+#[cfg(rings_native)]
+#[tokio::test]
+async fn test_restarted_exit_rejects_old_epoch_before_replay_state_or_side_effect() {
+    let client = session();
```
Comment (RyanKung) [inline crates/node/src/onion/circuit/tests/test_circuit_protocol.rs:1010]: need a comment about how this test works and why

File: crates/node/src/onion/circuit/crypto.rs [hunk 4]
```diff
@@ -203,17 +205,26 @@ pub async fn send_backward(
 fn build_forward_layers(
     client: OnionClientReturn,
     hops: &[OnionRouteHop],
+    process_epoch: crate::onion::OnionExitEpoch,
```
Comment (RyanKung) [inline crates/node/src/onion/circuit/crypto.rs]: `use crate::onion::OnionExitEpoch` , import first, never use crate::type in a fn sig.

File: crates/node/src/onion/circuit/mod.rs [hunk 5]
```diff
@@ -91,7 +91,7 @@ pub(super) const MAX_ONION_CRYPTO_OPS_GLOBAL_PER_WINDOW: u32 = 8192;
 pub(super) const MAX_ONION_CRYPTO_BYTES_PER_WINDOW: u64 = 256 * 1024 * 1024;
 pub(super) const MAX_ONION_CRYPTO_BYTES_GLOBAL_PER_WINDOW: u64 = 512 * 1024 * 1024;
 pub(super) const MAX_ONION_CRYPTO_PEERS: usize = 64;
-pub(super) const ONION_AEAD_NAMESPACE: &str = "rings-node:onion-circuit:v1";
+pub(super) const ONION_AEAD_NAMESPACE: &str = "rings-node:onion-circuit:v2";
```
Comment (RyanKung) [inline crates/node/src/onion/circuit/mod.rs]: We don’t guarantee backward compatibility before v1.0.

File: crates/node/src/onion/route/tests/test_route_selection.rs [hunk 6]
```diff
@@ -56,6 +56,7 @@ fn signed_exit_for_session_network_at(
                 .account_verification_pubkey()
                 .map_err(Error::CoreError)?,
             session_public_key: session_sk.session_public_key(),
+            process_epoch: crate::onion::OnionExitEpoch::new([29; 16]),
```
Comment (RyanKung) [inline crates/node/src/onion/route/tests/test_route_selection.rs]: need fix, same reason

File: crates/node/src/onion/tests/test_exit_registry.rs [hunk 7]
```diff
@@ -201,6 +202,17 @@ fn test_exit_descriptor_signature_covers_schema_version() -> Result<()> {
     Ok(())
 }
 
+#[test]
+fn test_exit_descriptor_signature_covers_process_epoch() -> Result<()> {
```
Comment (RyanKung) [inline crates/node/src/onion/tests/test_exit_registry.rs:213]: need description

**Review by RyanKung — CHANGES_REQUESTED** [review]

File: crates/node/src/onion/circuit/crypto.rs (hunk 4 above)
Comment (r2d2-rs) [reply crates/node/src/onion/circuit/crypto.rs]: Fixed in b44930ee. Imported `OnionExitEpoch` and use the short type in both function signatures.

File: crates/node/src/onion/circuit/tests/test_circuit_protocol.rs (hunk 2 above)
Comment (r2d2-rs) [reply crates/node/src/onion/circuit/tests/test_circuit_protocol.rs]: Fixed in b44930ee. Replaced the literal with documented `RESTART_REPLAY_CIRCUIT_ID`, which keeps this replay witness distinct from neighboring fixtures.

File: crates/node/src/onion/tests/test_exit_registry.rs (hunk 7 above)
Comment (r2d2-rs) [reply crates/node/src/onion/tests/test_exit_registry.rs:213]: Fixed in b44930ee. Added a description of the signature-substitution property the test witnesses.

File: crates/node/src/onion/route/tests/test_route_selection.rs (hunk 6 above)
Comment (r2d2-rs) [reply crates/node/src/onion/route/tests/test_route_selection.rs]: Fixed in b44930ee. Imported `OnionExitEpoch` and replaced `[29; 16]` with documented `TEST_EXIT_PROCESS_EPOCH`.

File: crates/node/src/onion/circuit/tests/test_circuit_protocol.rs (hunk 1 above)
Comment (r2d2-rs) [reply crates/node/src/onion/circuit/tests/test_circuit_protocol.rs]: Fixed in b44930ee. Replaced the literal with documented `TEST_PROCESS_EPOCH`; the concrete non-zero bytes only make the fixture deterministic.

File: crates/node/src/onion/circuit/tests/test_circuit_protocol.rs (hunk 3 above)
Comment (r2d2-rs) [reply crates/node/src/onion/circuit/tests/test_circuit_protocol.rs:1010]: Fixed in b44930ee. Added a test contract explaining the same `SessionSk` plus empty replay cache plus new process epoch, and why rejection must happen before an exit effect.

File: crates/node/src/onion/circuit/mod.rs (hunk 5 above)
Comment (r2d2-rs) [reply crates/node/src/onion/circuit/mod.rs]: Fixed in b44930ee. Restored the v1 AEAD namespace. The per-process epoch remains the restart replay boundary.

File: crates/node/src/onion/circuit/tests/test_circuit_protocol.rs
Comment (r2d2-rs) [reply crates/node/src/onion/circuit/tests/test_circuit_protocol.rs:1]: Fixed in b44930ee. Moved the native cfg to the circuit test module boundary and removed the redundant item-level gates.

File: crates/node/src/onion/mod.rs [hunk 8]
```diff
@@ -772,11 +798,30 @@ impl OnionExitRegistration {
...
         services: Vec<OnionExitService>,
         policy: OnionExitPolicy,
+    ) -> Self {
+        Self::with_process_epoch(
+            heartbeat_interval,
+            ttl,
+            node_type,
+            services,
+            policy,
+            OnionExitEpoch::random(),
```
[the remark this answers is not in the document]
Comment (r2d2-rs) [reply crates/node/src/onion/mod.rs]: Fixed in 85355739. `OnionExitRegistration` and its constructor are now crate-internal, the unsafe prelude export is removed, and `ProcessorBuilder` remains the sole constructor path that injects the same epoch into descriptor publication and circuit admission.

**Review by RyanKung — APPROVED** [review]
LGTM
